### PR TITLE
Updates to zend-expressive-router 3.0.0alpha2

### PR DESCRIPTION
**Pull request by weierophinney**
zend-expressive-router 3.0.0alpha2 renames middleware it previously defined, defines new middleware, and creates a number of new package-level constants for virtual services.

This patch:

- **Removes:**
  - the `RouteMiddlewareFactory` and `DispatchMiddlewareFactory`, as these are now provided by the zend-expressive-router package, and registered to the container by that package's config provider.
  - the `Implicit*Middleware` classes and related factories; these are now provided by the zend-expressive-router package.
- **Renames:**
  - the `Middleware\NotFoundMiddleware` to `Handler\NotFoundHandler`, and renames its factory accordingly. The class was acting as a handler, and can be used in more contexts as a handler than as middleware.
- **Adds:**
  - `Zend\Expressive\Container\StreamFactory`, for producing empty, writable stream instances.
  - Package-level constants by which virtual services may be referenced. In all cases, these refer to the FQCN variants previously used to ensure the old values continue to resolve; however, users are encouraged to use the new constants instead, as they will resolve, and because their docblocks will provide context for usage.
- **Changes:**
  - The `ConfigAggregator` was updated to use the new package-level constants, re-map the various aliases to the new services they refer to, and to add entries for various zend-expressive-router virtual services.
  - The `ApplicationConfigInjectionDelegator` now raises an exception if the callback provided to it does not produce an `Application` instance.
  - The `ApplicationConfigInjectionDelegator` changes on exception message to remove details that are not relevant in the 3.0 branch.

These should likely be the last breaking changes for this package before the 3.0.0 release.

### CODE REVIEW

File: src/Container/ApplicationConfigInjectionDelegator.php [hunk 1]
```diff
@@ -23,15 +24,20 @@ class ApplicationConfigInjectionDelegator
...
+     *     something other than an `Application` instance, as the delegator cannot
+     *     proceed with its operations.
      */
-    public function __invoke(ContainerInterface $container, string $serviceName, callable $callback)
+    public function __invoke(ContainerInterface $container, string $serviceName, callable $callback) : Application
     {
         $application = $callback();
         if (! $application instanceof Application) {
-            return $application;
+            throw new Exception\InvalidServiceException(sprintf(
```
Comment (webimpress) [inline src/Container/ApplicationConfigInjectionDelegator.php]: Why not static method in exception?

File: src/Container/ApplicationConfigInjectionDelegator.php [hunk 2]
```diff
@@ -65,8 +71,9 @@ public function __invoke(ContainerInterface $container, string $serviceName, cal
      *     'middleware_pipeline' => [
      *         // An array of middleware to register with the pipeline.
      *         // entries to register prior to routing/dispatching...
-     *         // - entry for \Zend\Expressive\Router\PathBasedRoutingMiddleware::class
-     *         // - entry \Zend\Expressive\Router\DispatchMiddleware::class
+     *         // - entry for \Zend\Expressive\Router\Middleware\PathBasedRoutingMiddleware::class
+     *         // - entry for \Zend\Expressive\Router\Middleware\MethodNotAllowedMiddleware::class
+     *         // - entry \Zend\Expressive\Router\Middleware\DispatchMiddleware::class
```
Comment (webimpress) [inline src/Container/ApplicationConfigInjectionDelegator.php]: For consistency should we have `entry for...` ?

File: src/constants.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+const ROUTE_MIDDLEWARE = __NAMESPACE__ . '\Middleware\RouteMiddleware';
+
+/**
+ * Virtual service name that should resolve to a service returning a response
+ * based on a `Throwable` argument produced when generating the application
+ * request.
+ *
+ * @var string
+ */
+const SERVER_REQUEST_ERROR_RESPONSE_GENERATOR = __NAMESPACE__ . '\Expressive\ServerRequestErrorResponseGenerator';
```
Comment (webimpress) [inline src/constants.php]: `__NAMESPACE__` is `Zend\Expressive`, should we have `\Expressive` part twice here?

File: src/constants.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * Virtual service name that should resolve to a service capable of producing
+ * a PSR-7 ServerRequestInterface instance for the application.
+ *
+ * @var string
+ */
+const SERVER_REQUEST_FACTORY = __NAMESPACE__ . '\Expressive\ServerRequestFactory';
```
Comment (webimpress) [inline src/constants.php]: As above.

File: test/Router/IntegrationTest.php [hunk 5]
```diff
@@ -103,6 +106,7 @@ private function createApplicationWithGetPost($adapter, $getName = null, $postNa
         $router = new $adapter();
         $app = $this->createApplicationFromRouter($router);
         $app->pipe(new RouteMiddleware($router, $this->response));
```
Comment (webimpress) [inline test/Router/IntegrationTest.php]: Here and in many other usages - the second parameter should be remove, `RouteMiddleware::__construct` has now one parameter.

File: src/Container/ApplicationConfigInjectionDelegator.php (hunk 1 above)
Comment (weierophinney) [reply src/Container/ApplicationConfigInjectionDelegator.php]: This one hasn't had one defined ever. I can add some later, but in most cases, we'll have different messages for every single factory, making it less viable.

File: src/constants.php (hunk 3 above)
Comment (weierophinney) [reply src/constants.php]: Good catch; fixing!

File: test/Router/IntegrationTest.php (hunk 5 above)
Comment (weierophinney) [reply test/Router/IntegrationTest.php]: Another nice catch! Fixing!

**Pull request merged**